### ISSUE TITLE
How to validate against json schema containing circular $ref?

### ISSUE THREAD

**Issue by csimplestring** [justinrainbow/json-schema#274]
Hi, thanks for this great work. I found that there is a new major version released to support $ref. However, where can I find the usage example about it?

Now I have json schema files, which some of them contains circular **$ref**. Namely, _schema-A.json_ has a **$ref** to _schema-B.json_, _schema-B.json_ has a **$ref** to _schema-A.json_.

Any help is appreciated and thanks in advanced.

**Comment by moderndeveloperllc** [comment]
For an example of `$ref` usage, I would take a look here: http://json-schema.org/example2.html. Circular `$ref`, like circular references, or arguments for that matter, will probably cause issues. The idea behind using `$ref` is to have reusable "modules" of information that can be used elsewhere. A circular reference would mean that you are referencing a parent object within a child object?

**Comment by jojo1981** [comment]
@csimplestring The schema can have circular dependencies which are resolved by the **RefResolver**. It will create an object graph in the memory. This schema in memory can be used to validate your data against. So this library can handle all **$ref** cases.

**Comment by csimplestring** [comment]
Can you give me an example code? Thanks!

---Sent from Boxer | http://getboxer.com

On 28 May 2016 09:56:56 CEST, Joost Nijhuis <notifications@github.com> wrote:@csimplestring The schema can have circular dependencies which are resolved by the RefResolver. It will create an object graph in the memory. This schema in memory can be used to validate your data against. So this library can handle all $ref cases.  —You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by csimplestring** [comment]
@jojo1981 Thanks for your work. I just try it by using it. When I want to validate against a circular reference json schema, it seems to fail due to infinite recursion loop and use up all memory. Do I use the RefResolver in a wrong way? 

I just use the RefResolver as the example in readme, also I notice there is a big change in API. So I guess maybe it is better to give me a demonstration or example in a correct way?

**Comment by psafarov** [comment]
Yes, **RefResolver** inlines all refs, so you get an infinite loop if you have recursion in your schema

**Comment by psafarov** [comment]
It's necessary to rethink schema compilation to avoid this problem

**Comment by jojo1981** [comment]
@csimplestring and @psafarov The **RefResolver** should be able to handle this. Can verify you're using the newest version of this library? Also an you add the schema you're using? Then I can try to reproduce this.

**Comment by psafarov** [comment]
@jojo1981 unfortunately, I cant publish the schema I used, but what I found is that it occurs because object with reference to itself inside gets to ObjectIterator which iterate object recursively

**Comment by psafarov** [comment]
#277 the issue is solved in this pull request

**Comment by csimplestring** [comment]
Hi, @jojo1981  Sorry for late response. After checking with my colleages, we found that the RefResovler should not be re-used in a for-loop.

In our code, we want to cache all the resolved schemas, so we write like this:
```
// This will cause infinite looping and crash
$refResolver = new RefResolver(new UriRretriever(), new UriResolver());
for ($schemas as $schema) {
  $resolved = $refResolver->resolver($schema);
  $cache[] = $resolved;
}

// but this will work ok
for ($schemas as $schema) {
  $refResolver = new RefResolver(new UriRretriever(), new UriResolver());
  $resolved = $refResolver->resolver($schema);
  $cache[] = $resolved;
}
```
I guess the RefResolver maintains a cache internally, could you explain a bit?

**Comment by bighappyface** [comment]
Circular references came in #277 and are available in 3.0.0
